csrc/aes_cbc.cpp

```diff
@@ -136,6 +152,68 @@ class AesCbcCipher {
         return result;
     }
 
+    int extended_update(bool is_iso10126_padding,
+        bool is_enc,
+        jbyteArray j_last_block,
+        uint8_t const* input,
+        int input_len,
+        uint8_t* output,
+        int unprocessed_input)
+    {
+        if (!is_iso10126_padding || is_enc) {
+            return update(input, input_len, output, unprocessed_input);
+        }
+        // We are decrypting and padding is ISO10126. In this case, we must not decrypt the last block until do_final.
+        // The last_block is used to store this information.
+        JBinaryBlob last_block(jenv_, nullptr, j_last_block);
+        // The last byte of last_block stores the number of bytes in it from the previous call.
+        int last_block_len = last_block.get()[AES_CBC_BLOCK_SIZE_IN_BYTES];
+        // This is the total number of bytes that we have. Some must be stored in the last_block and the rest must be
+        // passed to the cipher.
+        int all = last_block_len + input_len;
+        int rem = all % AES_CBC_BLOCK_SIZE_IN_BYTES;
+        // We need to store no more than 16 bytes of all the input. c holds the number of bytes that must be copied into
+        // last_block.
+        int c = rem == 0 ? min(AES_CBC_BLOCK_SIZE_IN_BYTES, all) : rem;
+        // p is the total number of bytes that must be passed to the cipher.
+        int p = all - c;
+        // p1 is the number of bytes that must be pass to the cipher from last_block.
+        int p1 = min(p, last_block_len);
+        // p2 is the number of bytes that must be pass to the cipher from input.
+        int p2 = p - p1;
+        // c2 is the number of bytes that must be copied from input to last_block.
+        int c2 = min(c, input_len);
+        // c1 is the number of bytes that must be copied from last_block to itself.
+        int c1 = c - c2;
+
+        // Processing step: passing bytes to the cipher. This includes passing bytes from last_block and input:
+        // Passing bytes from last_block:
+        int up = unprocessed_input - last_block_len;
+        int result = update(last_block.get(), p1, output, up);
+        up = (p1 + up) - result;
+
+        // Passing bytes from input
+        result += update(input, p2, output + result, up);
+
+        // Copying step: storing bytes into last_byte. This includes copy from last_block into itself and from input
+        // to last_block.
+        // Copy c1 bytes from last_block to itself.
+        int from_index = last_block_len - c1;
+        for (int i = 0; i < c1; i++) {
+            last_block.get()[i] = last_block.get()[from_index + i];
+        }
+        // Copy c2 from input to last_block.
+        from_index = input_len - c2;
+        for (int i = 0; i < c2; i++) {
+            last_block.get()[c1 + i] = input[from_index + i];
+        }
+        // Record the new size of last_block.
+        // Since c <= 16, the following cast is fine.
+        last_block.get()[AES_CBC_BLOCK_SIZE_IN_BYTES] = (uint8_t)c;
```

Comment:
I am not following how the last byte in the last_block is guaranteed to not be data. It seems like you're always holding onto the modulo of the last block (safe) or the last complete block in the case that we're aligned (possible data loss).

Reply:
Oh... Now I get it, are we using this extra byte at the end of this last-block array as a secret variable that stores the current length stored in the intermediate last-block scratch buffer? That state really needs to be tracked elsewhere, that's really unclear.

Reply:
I hope the added comments help now.

Reply:
The comments help, but I'd still prefer this tracked elsewhere like as a class member and/or separate variable if possible. This is really unconventional and liable to turn into nasty bugs in the future. I won't push further than this and I'll defer to the rest of the team for whether we want to push for a refactor.

Reply:
It can be passed as a separate variable but it needs to be wrapped in something like an array. This value is read and *written* to. Having it as an integer or a primitive type that gets passed from Java to C++ is not going to work.